log4j-core/src/main/java/org/apache/logging/log4j/core/layout/AbstractStringLayout.java

```diff
@@ -134,7 +134,13 @@ private static boolean isPreJava8() {
         final String version = System.getProperty("java.version");
```

Comment:
Can we make this method delegate to a package-private `static boolean isPreJava8(String javaVersion)` which we can invoke directly from the test? The reflective method is likely to cause failures if we refactor this code in the future, and setting the `java.version` system property may have other consequences which I'd prefer to avoid.

Reply:
I agree, but wanted to make the code change as small as possible, since it is a very minor change.  

If we want to be more adventures, we should move the org.apache.logging.log4j.util.Constants#getMajorVersion to core, and just have everything call that.  Then the version detection is centralised and can we can test it as a public method.

How to you advise we move forward on this.

Reply:
Personally, I dislike checking for versions. I much prefer checking for features. But seeing that this was already implemented I wouldn't expect you to change that.

I am a bit confused by your question though. It looks like Constants has a getMajorVersion method that has the same problem. I would recommend you just fix that and then have AbstractStringLayout reference Constants.JAVA_MAJOR_VERSION

Reply:
Please see updated PR.  This is a cleaner implementation.